File: 0232-implement-queue-using-stacks/0232-implement-queue-using-stacks.cpp
class MyQueue {
public:
    stack<int> s1,s2;
    MyQueue() {
        
    }
    
    void push(int x) {
        s1.push(x);
    }
    
    int pop() {
        int num;
        while(!s1.empty()){
            s2.push(s1.top());
            s1.pop();
        }
        num = s2.top();
        s2.pop();
        while(!s2.empty()){
            s1.push(s2.top());
            s2.pop();
        }
        return num;
    }
    
    int peek() {
        int num;
        while(!s1.empty()){
            s2.push(s1.top());
            s1.pop();
        }
        num = s2.top();
        while(!s2.empty()){
            s1.push(s2.top());
            s2.pop();
        }
        return num;
    }
    
    bool empty() {
        return s1.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */

File: 0232-implement-queue-using-stacks/0232-implement-queue-using-stacks.cs
public class MyQueue {
    //  https://leetcode.com/problems/implement-queue-using-stacks/solutions/2917480/two-solutions-na-ve-optimized
    Stack<int> st = new();
    Stack<int> rev = new();
    public MyQueue() {
        
    }
    
    public void Push(int x) {
        Pump(from: rev, to: st);
        st.Push(x);
    }
    
    public int Pop() => GetFirst(isPop: true);
    
    public int Peek() => GetFirst(isPop: false);
    
    public bool Empty() => st.Count + rev.Count == 0;

    private int GetFirst(bool isPop){
        Pump(from: st, to: rev);
        return isPop ? rev.Pop() : rev.Peek();
    }

    //  from 스택에서 to 스택으로 from의 모든 요소를 옮기는 메서드
    private void Pump(Stack<int> from, Stack<int> to){
        while(from.Count > 0) to.Push(from.Pop());
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.Push(x);
 * int param_2 = obj.Pop();
 * int param_3 = obj.Peek();
 * bool param_4 = obj.Empty();
 */